README.md
# FPL2FMS
FPL2FMS is an app to convert Garmin FPL files - which is generated by SkyVector - to X-Plane 11 FMS files – which is used by X-Plane 11 GNS430, GNS530, G1000, and default FMS -.

## Installation
Go to releases and download the latest release. The pre-compiled versions are provided for Windows and Linux.

## Build
You need to have Qt 5 and a static version of [pugixml](https://pugixml.org/).
Create a folder named 'lib' and put the compiled library file in it. Create a folder named 'include' inside the 'lib' folder and put the include files in it.

### Windows:
Assuming Qt 5 binaries are in PATH, and using mingw32:

```
git clone https://github.com/abdullah-radwan/FPL2FMS.git FPL2FMS
cd FPL2FMS
qmake
mingw32-make
mingw32-make clean
```

### Linux:

```
git clone https://github.com/abdullah-radwan/FPL2FMS.git ~/FPL2FMS
cd ~/FPL2FMS
qmake
make
make clean
chmod u+x FPL2FMS
./FPL2FMS
```

## About
The program is based on Qt 5 and pugixml. It's under GPL v3 license.

Any suggestions and contributes are welcomed.

Copyright © Abdullah Radwan


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.1 - 2021-01-31
### Added
- Delete the FPL file after conversion option.
- The input file name is set as the default file name in the save dialog.
- The window size and position are saved.
### Changed
- Code formatting style.
- Upgraded to Qt 5.15.2 and pugixml 1.11.4.
### Fixed
- Desktop shortcut created incorrectly on Windows.

## 1.0 - 2020-09-28
Initial release.

src/main.cpp
#include "mainwindow.h"

#include <QApplication>
#include <QDateTime>
#include <QStandardPaths>
#include <QDir>
#include <QDebug>

static QFile* logFile;

void messageHandler(QtMsgType type, const QMessageLogContext &context, const QString &msg)
{
    Q_UNUSED(context) Q_UNUSED(type)

    QTextStream out(logFile);

    out << QDateTime::currentDateTime().toString("hh:mm:ss.zzz ");

    out << msg << Qt::endl;

    out.flush();
}

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
    app.setApplicationName("FPL2FMS");
    app.setApplicationDisplayName("FPL2FMS");
    app.setApplicationVersion("1.1");

    QString dirPath = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
    QDir(dirPath).mkpath(".");

    logFile = new QFile(dirPath + "/log.txt");
    logFile->open(QFile::WriteOnly | QFile::Text);

    qInstallMessageHandler(messageHandler);

    qDebug() << "FPL2FMS 1.1, 31 January 2021";
    qDebug() << "Current date:" << QDateTime::currentDateTime().toString("yyyy-MM-dd");

    MainWindow mainWindow(dirPath);

    int result = app.exec();

    logFile->close();
    delete logFile;

    return result;
}


src/mainwindow.h
#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include "converter.h"

#include <QMainWindow>
#include <QSettings>
#include <QThread>
#include <QtAutoUpdaterWidgets/UpdateController>

QT_BEGIN_NAMESPACE
namespace Ui { class MainWindow; }
QT_END_NAMESPACE

class MainWindow : public QMainWindow
{
    Q_OBJECT
public:
    MainWindow(QString dirPath);
    ~MainWindow();

signals:
    void convert();

public slots:
    void onConvertingFile();
    void onDone();
    void onError(QString error);

private slots:
    void on_fplButton_clicked();
    void on_convertButton_clicked();
    void on_aboutAction_triggered();
    void on_checkAction_triggered();

private:
    Ui::MainWindow* ui;
    QSettings settings;
    QtAutoUpdater::Updater* updater;

    QString inputFolder;
    QString outputFolder;
    bool checkUpdates;

    Converter* converter;
    QThread converterThread;

    void readConfig();
    void writeConfig();

    void setConverter();
    void setUpdater();
};
#endif // MAINWINDOW_H


src/converter.cpp
#include "converter.h"
#include "pugixml.hpp"

#include <QFile>
#include <QDate>
#include <QDebug>

void Converter::convert()
{
    emit convertingFile();

    waypoints.clear();

    qDebug() << "Reading" << data.inputPath;

    pugi::xml_document doc;
    pugi::xml_parse_result result = doc.load_file(data.inputPath.toStdString().c_str());

    if (result.status != pugi::status_ok)
    {
        qDebug() << "Error:" << result.description();
        emit error(result.description());
        return;
    }

    for (const auto& fplWaypoint : doc.child("flight-plan").child("waypoint-table").children("waypoint"))
    {
        Waypoint waypoint;

        waypoint.type = convertType(fplWaypoint.child_value("type"));
        waypoint.ident = fplWaypoint.child_value("identifier");
        waypoint.lat = fplWaypoint.child_value("lat");
        waypoint.lon = fplWaypoint.child_value("lon");

        waypoints.append(waypoint);
    }

    setRoute();
}

QString Converter::convertType(QString type)
{
    if (type == "AIRPORT") return "1";

    if (type == "NDB") return "2";

    if (type == "VOR") return "3";

    if (type == "INT") return "11";

    if (type == "USER WAYPOINT") return "28";

    return "0";
}

void Converter::setRoute()
{
    QString departure = waypoints.first().ident;
    QString arrival = waypoints.last().ident;

    waypoints.first().airway = "ADEP";
    waypoints.last().airway = "ADES";

    QStringList route = data.route.split(" ");

    for (auto& waypoint : route)
        if (waypoint.isEmpty()) route.removeOne(waypoint);

    qDebug() << "Route:" << route;

    for (auto& waypoint : waypoints)
    {
        if (waypoint.ident == departure || waypoint.ident == arrival) continue;

        if (route.empty())
        {
            qDebug() << "Invalid route.";
            emit error("the flight route is invalid.");
            return;
        }

        if (waypoint.ident == route.first())
        {
            waypoint.airway = "DRCT";
            route.removeFirst();

            continue;
        }

        if (route.size() < 2)
        {
            qDebug() << "Invalid route" << route;
            emit error("the flight route is invalid.");
            return;
        }

        if (waypoint.ident == route.at(1))
        {
            waypoint.airway = route.first();

            route.removeFirst();
            route.removeFirst();

            continue;
        }

        waypoint.airway = route.first();
    }

    writeFMS();
}

void Converter::writeFMS()
{
    qDebug() << "Writing into" << data.outputPath;
    QFile file(data.outputPath);

    if (!file.open(QIODevice::WriteOnly))
    {
        qDebug() << "Couldn't open the FMS file";
        emit error("couldn't create the FMS file.");
        return;
    }

    QTextStream out(&file);
    out << "I" << Qt::endl;
    out << "1100 Version" << Qt::endl;
    out << "CYCLE " << getCycle() << Qt::endl;
    out << "ADEP " << waypoints.first().ident << Qt::endl;
    out << "ADES " << waypoints.last().ident << Qt::endl;
    out << "NUMENR " << waypoints.size() << Qt::endl;

    for (auto& waypoint : waypoints)
        out << waypoint.type << " " << waypoint.ident << " " << waypoint.airway << " 0.000000 " << waypoint.lat << " " << waypoint.lon << Qt::endl;

    file.close();

    if (data.deleteFPL)
    {
        QFile fplFile(data.inputPath);

        if (fplFile.remove()) qDebug() << "The FPL file was deleted successfully";
        else qDebug() << "Couldn't delete the FPL file";
    }

    qDebug() << "The file converted was successfully";

    emit done();
}

QString Converter::getCycle()
{
    QDate currentDate = QDate::currentDate();
    QDate date = QDate::fromString("2021/01/28", "yyyy/MM/dd");

    int cycle = 1;
    int year = date.year();

    QMap<QDate, int> cycleMap;
    cycleMap.insert(date, cycle);

    while (true)
    {
        date = date.addDays(28);

        if (currentDate.daysTo(date) >= 0)
        {
            QString lastCycle;

            if (cycleMap.last() < 10) lastCycle = "0" + QString::number(cycleMap.last());
            else lastCycle = QString::number(cycleMap.last());

            QString sCycle = QString::number(cycleMap.lastKey().year()).remove(0, 2) + lastCycle;
            qDebug() << "Cycle:" << sCycle;
            return sCycle;
        }

        if (date.year() > year)
        {
            cycleMap.clear();
            cycle = 0;
            year = date.year();
        }

        cycle++;
        cycleMap.insert(date, cycle);
    }
}


src/mainwindow.cpp
#include "mainwindow.h"
#include "ui_mainwindow.h"

#include <QFileDialog>
#include <QMessageBox>
#include <QCloseEvent>

MainWindow::MainWindow(QString dirPath) : ui(new Ui::MainWindow), settings(dirPath + "/config.cfg", QSettings::IniFormat), converterThread(this)
{
    ui->setupUi(this);

    settings.setParent(this);
    readConfig();

    setConverter();
    setUpdater();

    show();

    if (checkUpdates) on_checkAction_triggered();
}

void MainWindow::readConfig()
{
    settings.beginGroup("Settings");

    inputFolder = settings.value("InputFolder").toString();
    outputFolder = settings.value("OutputFolder").toString();
    ui->deleteCheck->setChecked(settings.value("DeleteFPL", true).toBool());
    checkUpdates = settings.value("CheckDate", 0).toDate().addDays(14) <= QDate::currentDate();

    settings.endGroup();

    settings.beginGroup("MainWindow");

    QSize size = settings.value("Size", QSize(-1, -1)).toSize();
    if (size != QSize(-1, -1)) resize(size);

    QPoint pos = settings.value("Pos", QPoint(-1, -1)).toPoint();
    if (pos != QPoint(-1, -1)) move(pos);

    settings.endGroup();
}

void MainWindow::setConverter()
{
    converter = new Converter;

    converter->moveToThread(&converterThread);

    connect(this, &MainWindow::convert, converter, &Converter::convert);

    connect(converter, &Converter::convertingFile, this, &MainWindow::onConvertingFile);
    connect(converter, &Converter::done, this, &MainWindow::onDone);
    connect(converter, &Converter::error, this, &MainWindow::onError);

    connect(&converterThread, &QThread::finished, converter, &QObject::deleteLater);

    converterThread.start();
}

void MainWindow::setUpdater()
{
    updater = QtAutoUpdater::Updater::create("qtifw", {{"path", qApp->applicationDirPath() + "/maintenancetool"}}, qApp);

    if (updater)
    {
        connect(updater, &QtAutoUpdater::Updater::checkUpdatesDone, [&](QtAutoUpdater::Updater::State state)
        {
            switch (state)
            {
            case QtAutoUpdater::Updater::State::Error:
                qDebug() << "An error occured";
                checkUpdates = false;
                break;

            case QtAutoUpdater::Updater::State::NoUpdates:
                qDebug() << "No updates were found";
                break;

            case QtAutoUpdater::Updater::State::NewUpdates:
                qDebug() << "An update was found" << updater->updateInfo();
                break;

            default:
                break;
            }
        });
    }
    else qDebug() << "Couldn't create the updater";
}

void MainWindow::on_fplButton_clicked()
{
    QString path = QFileDialog::getOpenFileName(this, "Select FPL File", inputFolder, "FPL Files (*.fpl)");

    if (path.isEmpty()) return;

    inputFolder = QFileInfo(path).absolutePath();

    converter->data.inputPath = path;

    ui->fplLabel->setText("Selected file: " + path);
}

void MainWindow::on_convertButton_clicked()
{
    if (converter->data.inputPath.isEmpty())
    {
        statusBar()->showMessage("You must choose an FPL file.");
        return;
    }

    converter->data.route = ui->routeEdit->text();

    if (converter->data.route.isEmpty())
    {
        statusBar()->showMessage("The flight route can't be empty.");
        return;
    }

    QString savePath;

    if (outputFolder.isEmpty())
    {
        savePath = converter->data.inputPath;
        savePath.remove(".fpl");
        savePath.append(".fms");
    }
    else savePath = outputFolder + "/" + QFileInfo(converter->data.inputPath).baseName().append(".fms");

    QString path = QFileDialog::getSaveFileName(this, "Save FMS File", savePath, "FMS Files (*.fms)");

    if (path.isEmpty()) return;

    if (!path.endsWith(".fms", Qt::CaseInsensitive)) path.append(".fms");

    ui->convertButton->setEnabled(false);

    outputFolder = QFileInfo(path).absolutePath();

    converter->data.outputPath = path;

    converter->data.deleteFPL = ui->deleteCheck->isChecked();

    emit convert();
}

void MainWindow::onConvertingFile()
{
    statusBar()->showMessage("Converting the FPL file...");
}

void MainWindow::onDone()
{
    statusBar()->showMessage("The FPL file was converted successfully.");
    ui->convertButton->setEnabled(true);
}

void MainWindow::onError(QString error)
{
    statusBar()->showMessage("Error: " + error);
    ui->convertButton->setEnabled(true);
}

void MainWindow::on_checkAction_triggered()
{
    if (!updater)
    {
        QMessageBox::warning(this, "Updater initialization", "Couldn't initialize the updater.");
        return;
    }

    qDebug() << "Starting updater";

    (new QtAutoUpdater::UpdateController(updater, this))->start(QtAutoUpdater::UpdateController::DisplayLevel::ExtendedInfo);
}


void MainWindow::on_aboutAction_triggered()
{
    QMessageBox::about(this, "FPL2FMS", "FPL2FMS\n"
                                        "An app to convert Garmin FPL files to X-Plane 11 FMS files.\n\n"

                                        "Version: 1.1\n"
                                        "Release date: 31 January 2021\n\n"

                                        "Copyright © Abdullah Radwan");
}

void MainWindow::writeConfig()
{
    settings.beginGroup("Settings");

    settings.setValue("InputFolder", inputFolder);
    settings.setValue("OutputFolder", outputFolder);
    settings.setValue("DeleteFPL", ui->deleteCheck->isChecked());

    if (checkUpdates || !settings.value("CheckDate", 0).toDate().isValid()) settings.setValue("CheckDate", QDate::currentDate());

    settings.endGroup();

    settings.beginGroup("MainWindow");

    settings.setValue("Size", size());
    settings.setValue("Pos", pos());

    settings.endGroup();
}

MainWindow::~MainWindow()
{
    converterThread.quit();
    converterThread.wait();

    writeConfig();

    delete ui;
}
